### ISSUE TITLE
support more sophisticated fixme warnings

### ISSUE THREAD

**Issue by ahvigil** [PyCQA/pylint#2874]
### Is your feature request related to a problem? Please describe
pylint warns for configured fixme notes (defaulting to FIXME/XXX/TODO) but sometimes it would be desirable to be more flexible in how these are flagged. For example, `# FIXME: this is broken` by itself might not be ok but `# FIXME: this is broken (ISSUE-1234)` might be acceptable because it points to an externally tracked issue

### Describe the solution you'd like
The `notes` config value should optionally allow a regex pattern to handle more complicated `fixme` conditions (eg `FIXME*ISSUE-\d+`).

### Additional context
The checker for this rule already [uses a regex pattern](https://github.com/PyCQA/pylint/blob/master/pylint/checkers/misc.py#L108-L112) to flag configured notes so I think this should be easy to implement, but patterns are currently escaped before being inserted into the regex.

**Comment by PCManticore** [comment]
Sounds good, thanks for reporting the issue.

**Comment by BennyTheSen** [comment]
I will try to tackle this.
@PCManticore should this be done with an additional option like `notes-rgx` ? 
I think the proposed solution to include regex in the `notes` option could cause confusion if someone is trying to detect comments like `#./TODO` without escaping regex specific tokens.

**Comment by PCManticore** [comment]
@BennyTheSen Thanks for tackling this. Yes, a separate option would be the way to go. This maps nicely to the rest of our options that also accept a regex counterpart.

**Comment by stdedos** [comment]
_I'd say "Closed as done"?_

**Comment by stdedos** [comment]
... Although do you mind explaining how does it work though? 😕 

Assuming file:
```py
"""test""

__version__ = "0.1"

# FixMe: test
# FixMe: ABC-1: test-abc
# FixMe: CBA-1: test-cba
```

and 
```
notes-rgx=(FIXME|XXX|TODO)(?!.*ABC-\d+)
```

I expected:
![image](https://user-images.githubusercontent.com/133706/177769391-c08a26b4-6f37-4648-a70c-7364fe277674.png)

but I got:
```console
$  pylint src/__version__.py 
************* Module __version__
src/__version__.py:5:1: W0511: FixMe: test (fixme)
src/__version__.py:6:1: W0511: FixMe: ABC-1: test-abc (fixme)
src/__version__.py:7:1: W0511: FixMe: CBA-1: test-cba (fixme)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

"In this case" I'd like, apart from the existing functionality, a `notes-rgx-ignore` regex (i.e. drop any `notes` that _also matches_ `notes-rgx-ignore`).

As you can appreciate, writing negative regex is not the funniest job on the planet - especially if the regex starts getting complicated.

_Ofc, feel free to update documentation too ..._

**Comment by DanielNoord** [comment]
Basically we warn whenever we find a pattern that matches the regex.
<img width="761" alt="Schermafbeelding 2022-07-07 om 14 20 50" src="https://user-images.githubusercontent.com/13665637/177771804-1aa3a50e-1988-43d8-8c64-34cde32eac6d.png">

The above is an example of how I use this in one of my own projects. I want to allow `TODO` if there is an issue number associated with it, otherwise I want to flag.

I don't think we'll add an `-ignore` option. I know `regex` can be difficult, but I don't think there is anything that the current system does not allow you to do with a little regex crafting.

**Comment by stdedos** [comment]
Also
```ini

[MISCELLANEOUS]

# List of note tags to take in consideration, separated by a comma.
notes="" # Remove the default entries
```

for the original config file

**Comment by stdedos** [comment]
pylint confuses me anyway so much, "it doesn't matter" 😓 

I am using the `.pylintrc` file, "maybe sometime" I am planning to move it to `pyproject.toml` (if available)

**Comment by DanielNoord** [comment]
> pylint confuses me anyway so much, "it doesn't matter" 😓

We're always happy to help with any issues 😄 The configurability of `pylint` has its downsides, making it a bit difficult to set up according to your wishes.

> I am using the `.pylintrc` file, "maybe sometime" I am planning to move it to `pyproject.toml` (if available)

That is indeed available! You an even use the `pylint-config` command to make the file right now and take your current configuration with you!

**Comment by stdedos** [comment]
Wuuuut? pylint has "auto-fix" things? 🤯 💥 (~~fix~~ > migrate)

Thanks for letting me know 😄

**Comment by DanielNoord** [comment]
> Wuuuut? pylint has "auto-fix" things? 🤯 💥 (~fix~ > migrate)

Sorry to disappoint! I meant the `configuration file`. By using `pylint-config` you can generate a new configuration file based on your current configuration. That way you can easily upgrade from an `pylintrc` to an `pyproject.toml` file.
